Fully native DTensor.__new__

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163030
* #162968
* __->__ #162508
* #161695

Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

swolchok committed:
Fully native DTensor.__new__

Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -928,6 +864,109 @@ static bool intern_dtensor_strings() {
+    extra_dispatch_keys = extra_dispatch_keys.add(c10::DispatchKey::Negative);
+  }
+
+  py::handle spec = py::handle(r.pyobject(2));
```
Does .pyobject() return a new or borrowed reference?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -928,6 +864,109 @@ static bool intern_dtensor_strings() {
+  }
+
+  py::handle spec = py::handle(r.pyobject(2));
+  const auto tensor_meta = spec.attr(dtensor_interned_strings.tensor_meta);
```
Some error checking from the old code is lost here when this is None?
Also what happens here when the attribute doesn't exist?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -928,6 +864,109 @@ static bool intern_dtensor_strings() {
+          /*allow_preexisting_pyobj=*/false,
+          // we know DTensor has __torch_dispatch__; avoid checking again.
+          /*has_torch_dispatch_if_known=*/true));
+  py_tensor.attr(dtensor_interned_strings._spec) = spec;
+  py_tensor.attr(dtensor_interned_strings._local_tensor) = local_tensor;
```
I've never seen this before. Is the refcounting acutally correct on this?!

albanD reviewed (commented):
Not really sure what the end goal here is and the level of optimization we're aiming for. But pybind is most likely costing you a lot, especially for things like setting a slot attribute?

swolchok committed:
rebase on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Python 3.13 renamed _PyCFunctionFast to remove the underscore on "Fully native DTensor.__new__"


Move the entirety of `__new__` into C++, saving a layer of disable_dynamo and making progress toward all-C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge
